### ISSUE TITLE
Very high memory usage in remove_adapter.py

### ISSUE THREAD

**Issue by holtgrewe** [ndaniel/fusioncatcher#19]
I'm seeing a very high memory usage in an RNA-seq data sets (2x151bp) of ~135M reads in `remove_adapter.py` (I'm running with 16 threads) -- 37GB and counting. As adapter trimming essentially requires O(1) memory, I assume that this is a bug (maybe a memory "leak" by never releasing any of the buffers?)

**Comment by ndaniel** [comment]
It is challenging to figure out what is going on in this case (e.g. what FusionCatcher version, does `remove_adapter.py` finish successfully, etc.).
One way around it in case that the memory is an issue, it would be to decrease the number of threads which would decrease the memory requirements.

**Comment by holtgrewe** [comment]
Would it make sense to add a setting to only enable multithreading in Bowtie, BLAT etc. but disabling in the adapter removal and other Python multiprocessing programs that show problems with multithreading?

**Comment by ndaniel** [comment]
Would be possible to have:
- FusionCatcher version, and
- Python version,
where this issue has been noticed?

**Comment by holtgrewe** [comment]
Sorry, sure:

- Python 2.7.9
- fusioncatcher 0.99.6a

**Comment by ndaniel** [comment]
I tried to reproduce the issue but I was not able to.

Running `/usr/bin/time -v remove_adapter.py ....` with 32 threads on 151 bp reads shows this in my case:
```
	User time (seconds): 367.51
	System time (seconds): 25.80
	Percent of CPU this job got: 158%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 4:08.87
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 5350272
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 6671098
	Voluntary context switches: 975607
	Involuntary context switches: 39014
	Swaps: 0
	File system inputs: 8
	File system outputs: 5382200
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

Another comment would that `remove_adapter.py` will use *only and only one* thread in most of the situations even that at command line is specified that more than one thread should be use. This is by design because the script will estimate depending on adapters found if it is faster to proceed with one thread or with many threads. In many cases there is no benefit in using several parallel threads.

The script `remove_adapter.py` indeed removes the adapters from the input FASTQ files but it is not use the **naive** approach of looking a pre-defined string (that is given by the user). `remove_adapter.py` is trying to remove *automatically* the adapter without *a priori* knowledge of the adapter string and by mapping (like mapping in BLAST) reads on each other. Therefore the memory requirement most likely is not O(1).

**Comment by ndaniel** [comment]
After testing several more FASTQ files I was able to reproduce the issue. In my tests this issues appears **only and only** in `remove_adapter.py` script and no other Python scripts.

This has been fixed and back ported into v0.99.6a.

**Comment by holtgrewe** [comment]
Thanks a lot. I like! :+1: :tada:
